src/Microsoft.Data/ArrowStringColumn.cs

```diff
@@ -0,0 +1,348 @@
+
+        public override BaseColumn Sort(bool ascending = true) => throw new NotSupportedException();
+
+        public override BaseColumn Clone(BaseColumn mapIndices = null, bool invertMapIndices = false, long numberOfNullsToAppend = 0)
+        {
+            ArrowStringColumn clone;
+            if (!(mapIndices is null))
+            {
+                if (mapIndices.DataType != typeof(long))
+                    throw new ArgumentException(Strings.MismatchedValueType + " PrimitiveColumn<long>", nameof(mapIndices));
+                if (mapIndices.Length > Length)
+                    throw new ArgumentException(Strings.MapIndicesExceedsColumnLenth, nameof(mapIndices));
+                clone = Clone(mapIndices as PrimitiveColumn<long>, invertMapIndices);
+            }
+            else
+            {
+                clone = Clone();
+            }
+            for (long i = 0; i < numberOfNullsToAppend; i++)
+            {
+                clone.Append(null);
+            }
+            return clone;
+        }
+
+        private ArrowStringColumn Clone(PrimitiveColumn<long> mapIndices = null, bool invertMapIndex = false)
+        {
+            ArrowStringColumn ret = new ArrowStringColumn(Name);
+            if (mapIndices is null)
+            {
+                for (long i = 0; i < Length; i++)
+                {
+                    ret.Append(this[i]);
+                }
+            }
+            else
+            {
+                if (mapIndices.Length > Length)
+                    throw new ArgumentException(Strings.MapIndicesExceedsColumnLenth, nameof(mapIndices));
+                if (invertMapIndex == false)
+                {
+                    for (long i = 0; i < mapIndices.Length; i++)
+                    {
+                        ret.Append(this[(long)mapIndices[i]]);
+                    }
+                }
+                else
+                {
+                    long mapIndicesLengthIndex = mapIndices.Length - 1;
+                    for (long i = 0; i < mapIndices.Length; i++)
+                    {
+                        ret.Append(this[(long)mapIndices[mapIndicesLengthIndex - i]]);
+                    }
+                }
+            }
+            return ret;
+        }
+
+        public Dictionary<TKey, ICollection<long>> HashColumnValues<TKey>()
+        {
+            if (typeof(TKey) == typeof(string))
+            {
+                Dictionary<string, ICollection<long>> multimap = new Dictionary<string, ICollection<long>>(EqualityComparer<string>.Default);
+                for (long i = 0; i < Length; i++)
+                {
+                    bool containsKey = multimap.TryGetValue(this[i] ?? default, out ICollection<long> values);
+                    if (containsKey)
+                    {
+                        values.Add(i);
+                    }
+                    else
+                    {
+                        multimap.Add(this[i] ?? default, new List<long>() { i });
```

Comment:
Doing `this[i]` is actually a pretty expensive operation, right? It will decode the UTF8 bytes into a UTF16 .NET string. We should probably only do it once for each index.

Reply:
Yup. Fixed. Added a temp patch to fix a crash in the process too that I overlooked. Added a comment to https://github.com/dotnet/corefxlab/issues/2700